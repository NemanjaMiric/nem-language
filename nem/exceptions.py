"""Raise exceptions.

Holds multiple exceptions.

Exceptions:
LexerException - During lexing
ParserException - During parsing
EvaluatorException - During evaluation

"""


class LexerException(Exception):

    """Raise exception.

    Gets raised when an error appears during lexing.

    """

    pass


class ParserException(Exception):

    """Raise exception.

    Gets raised when an error appears during parsing.

    """

    pass


class EvaluatorException(Exception):

    """Raise exception.

    Gets raised when an error appears during evaluation.

    """

    pass