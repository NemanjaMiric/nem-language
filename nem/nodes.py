"""Represent node elements.

Used for representing each node element for the AST.

"""


class Node:

    """Hold template.

    Template for node elements.

    """

    def __eq__(self, other):
        """Compare equality of classes."""
        if not isinstance(other, Node):
            raise NotImplementedError
        return type(self) == type(other) and self.__dict__ == other.__dict__


class Number(Node):

    """Hold a number.

    Holds a number.

    """

    def __init__(self, value):
        """Initialize Number class."""
        self.value = value

    def __repr__(self):
        """Represent Number class."""
        return "Number({})".format(repr(self.value))


class Variable(Node):

    """Hold a variable.

    Holds a variable.

    """

    def __init__(self, variable):
        """Initialize Variable class."""
        self.variable = variable

    def __repr__(self):
        """Represent Variable class."""
        return "Variable({})".format(repr(self.variable))


class Text(Node):

    """Hold text.

    Holds text.

    """

    def __init__(self, text):
        """Initialize Text class."""
        self.text = text

    def __repr__(self):
        """Represent Text class."""
        return "Text({})".format(repr(self.text))


class List(Node):

    """Hold a list.

    Holds a list.

    """

    def __init__(self, elements):
        """Initialize List class."""
        self.elements = elements

    def __repr__(self):
        """Represent List class."""
        return "List({})".format(repr(self.elements))


class ListIndex(Node):

    """Hold a list index.

    Holds a list index.

    """

    def __init__(self, holder, index):
        """Initialize ListIndex class."""
        self.holder = holder
        self.index = index

    def __repr__(self):
        """Represent ListIndex class."""
        return "ListIndex({}, {})".format(repr(self.holder), repr(self.index))


class Expressions(Node):

    """Hold expressions.

    Holds expressions.

    """

    def __init__(self, expressions):
        """Initialize Expressions class."""
        self.expressions = expressions

    def __repr__(self):
        """Represent Expressions class."""
        return "Expressions({})".format(repr(self.expressions))


class UnaryOperation(Node):

    """Hold a unary operation.

    Holds a unary operation.

    """

    def __init__(self, node, operator):
        """Initialize UnaryOperation class."""
        self.node = node
        self.operator = operator

    def __repr__(self):
        """Represent UnaryOperation class."""
        return "UnaryOperation({}, {})".format(repr(self.node), repr(self.operator))


class BinaryOperation(Node):

    """Hold a binary operation.

    Holds a binary operation.

    """

    def __init__(self, left_node, operator, right_node):
        """Initialize BinaryOperation class."""
        self.left_node = left_node
        self.operator = operator
        self.right_node = right_node

    def __repr__(self):
        """Represent BinaryOperation class."""
        return "BinaryOperation({}, {}, {})".format(repr(self.left_node), repr(self.operator), repr(self.right_node))


class AssignmentOperation(Node):

    """Hold an assignment operation.

    Holds an assignment operation.

    """

    def __init__(self, variable, value):
        """Initialize AssignmentOperation class."""
        self.variable = variable
        self.value = value

    def __repr__(self):
        """Represent AssignmentOperation class."""
        return "AssignmentOperation({}, {})".format(repr(self.variable), repr(self.value))


class IfElse(Node):

    """Hold an if-else statement.

    Holds an if-else statement.

    """

    def __init__(self, condition, if_expression, else_expression=None):
        """Initialize IfElse class."""
        self.condition = condition
        self.if_expression = if_expression
        self.else_expression = else_expression

    def __repr__(self):
        """Represent IfElse class."""
        return "IfElse({}, {}, {})".format(repr(self.condition), repr(self.if_expression), repr(self.else_expression))


class While(Node):

    """Hold a while loop.

    Holds an while loop.

    """

    def __init__(self, condition, expression):
        """Initialize While class."""
        self.condition = condition
        self.expression = expression

    def __repr__(self):
        """Represent While class."""
        return "While({}, {})".format(repr(self.condition), repr(self.expression))


class FunctionDefinition(Node):

    """Hold a function definition.

    Holds a function definition.

    """

    def __init__(self, name, parameters, expression):
        """Initialize FunctionDefinition class."""
        self.name = name
        self.parameters = parameters
        self.expression = expression

    def __repr__(self):
        """Represent FunctionDefinition class."""
        return "FunctionDefinition({}, {}, {})".format(repr(self.name), repr(self.parameters), repr(self.expression))


class FunctionCall(Node):

    """Hold a function call.

    Holds a function call.

    """

    def __init__(self, name, arguments):
        """Initialize FunctionCall class."""
        self.name = name
        self.arguments = arguments

    def __repr__(self):
        """Represent FunctionCall class."""
        return "FunctionCall({}, {})".format(repr(self.name), repr(self.arguments))


class Import(Node):

    """Hold an import.

    Holds an import.

    """

    def __init__(self, filename):
        """Initialize Import class."""
        self.filename = filename

    def __repr__(self):
        """Represent Import class."""
        return "Import({})".format(repr(self.filename))


class Return(Node):

    """Hold a return value.

    Holds a return value.

    """

    def __init__(self, value):
        """Initialize Return class."""
        self.value = value

    def __repr__(self):
        """Represent Return class."""
        return "Return({})".format(repr(self.value))


class Continue(Node):

    """Hold a continue.

    Holds a continue.

    """

    def __init__(self):
        """Initialize Continue class."""
        pass

    def __repr__(self):
        """Represent Continue class."""
        return "Continue()"


class Break(Node):

    """Hold a break.

    Holds a break.

    """

    def __init__(self):
        """Initialize Break class."""
        pass

    def __repr__(self):
        """Represent Break class."""
        return "Break()"


class Null(Node):

    """Hold null.

    Holds null.

    """

    def __init__(self):
        """Initialize Null class."""
        pass

    def __repr__(self):
        """Represent Null class."""
        return "Null()"
