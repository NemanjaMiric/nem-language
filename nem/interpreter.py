"""Interpret Nem code.

Holds class Interpreter and used for debugging the class Interpreter.

"""


from nem.nemtoken import Token
from nem.exceptions import LexerException, ParserException
import re


class Interpreter:

    """Represent the interpreter.

    Used for representing the interpreter as a class.

    """

    def __init__(self, code):
        """Initialize the Interpreter class.

        Initializes the Interpreter class and adds a newline character to the end of the Nem code for handling EOF.

        """
        self.code = "{}\n".format(code)

    def lex(self):
        """Lex the code.

        Lexer for Nem code.

        List of tokens:
        ',' - <list-element>
        ';' - <end>
        r'[=+-*/^%<>]' - <operator>
        '(' - <left-bracket>
        ')' - <right-bracket>
        '[' - <left-square>
        ']' - <right-square>
        '"..."' - <text>
        r'[0-9]*[.]?[0-9]+' - <number>
        r'[a-z_][a-z_0-9]*' - <symbol>
        'null' - <null>
        'input' - <keyword>
        'output' - <keyword>
        'error' - <keyword>
        'while' - <keyword>
        'if' - <keyword>
        'otherwise' - <keyword>
        'import' - <keyword>
        'open' - <keyword>
        'convert' - <keyword>
        'number' - <keyword>
        'text' - <keyword>
        'list' - <keyword>
        'break' - <keyword>
        'continue' - <keyword>
        'function' - <keyword>
        'not' - <operator>
        'is' - <operator>
        'and' - <operator>
        'or' - <operator>

        Gets ignored:
        '#...\n' - Comment

        """
        tokens = []
        current_index = 0

        line = 1
        character = 1

        while current_index < len(self.code):
            current_token = self.code[current_index]

            if current_token == "\n":
                line += 1
                character = 1

            elif current_token == ",":
                tokens.append(Token("list-element", current_token))

            elif current_token == ";":
                tokens.append(Token("end", current_token))

            elif current_token in "=+-*/^%<>":
                tokens.append(Token("operator", current_token))

            elif current_token == "[":
                tokens.append(Token("left-square", current_token))

            elif current_token == "]":
                tokens.append(Token("right-square", current_token))

            elif current_token == "(":
                tokens.append(Token("left-bracket", current_token))

            elif current_token == ")":
                tokens.append(Token("right-bracket", current_token))

            elif current_token == "#":
                while self.code[current_index] != "\n":
                    current_index += 1
                line += 1
                character = 0

            elif current_token == "\"":
                current_index += 1
                start_index = current_index
                try:
                    while self.code[current_index] != "\"":
                        current_index += 1
                        character += 1
                except IndexError:
                    raise LexerException("Lexing Error (Line {}, Character: {}): Text doesn't have an end"
                                         .format(line, character))
                else:
                    tokens.append(Token("text", self.code[start_index:current_index]))

            elif re.match("[a-zA-Z_]", current_token):
                start_index = current_index
                current_index += 1
                character += 1
                while re.match("[a-zA-Z_][a-zA-Z_0-9]*\Z", self.code[start_index:current_index + 1]):
                    current_index += 1
                    character += 1
                if self.code[start_index:current_index] in ("input", "output", "error", "while", "if",
                                                            "otherwise", "import", "open", "convert", "number",
                                                            "text", "list", "break", "continue", "function"
                                                            ):
                    tokens.append(Token("keyword", self.code[start_index:current_index]))
                elif self.code[start_index:current_index] == "null":
                    tokens.append(Token("null", self.code[start_index:current_index]))
                elif self.code[start_index:current_index] in ("not", "is", "and", "or"):
                    tokens.append(Token("operator", self.code[start_index:current_index]))
                else:
                    tokens.append(Token("symbol", self.code[start_index:current_index]))
                continue

            elif re.match("[0-9.]", current_token):
                start_index = current_index
                current_index += 1
                character += 1
                while re.match("[0-9]*[.]?[0-9]+\Z", self.code[start_index:current_index + 1])                         \
                        or self.code[start_index:current_index + 1] == ".":
                    current_index += 1
                    character += 1
                if self.code[start_index:current_index] == ".":
                    raise LexerException("Lexing Error (Line {}, Character: {}): Unrecognized token"
                                         .format(line, character))
                else:
                    tokens.append(Token("number", self.code[start_index:current_index]))
                continue

            elif current_token in " \t":
                pass

            else:
                raise LexerException("Lexing Error (Line {}, Character: {}): Unrecognized token"
                                     .format(line, character))

            current_index += 1
            character += 1

        return tokens

    def parse(self):
        """Parse the tokens.

        Parses the tokens that get returned from the lexer and builds an AST out of them.

        """
        pass


def main():
    """Debug the interpreter.

    Used as the entry-point when the file gets ran directly. It's usually used for debugging the class Interpreter.

    """
    # For Debug purposes
    while True:
        print(Interpreter(input(">> ")).lex())


if __name__ == "__main__":
    # Only activates when the file gets ran directly.
    main()
