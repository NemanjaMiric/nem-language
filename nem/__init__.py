"""Initialize modules."""

import nem.evaluator
import nem.exceptions
import nem.interpreter
import nem.lexer
import nem.nodes
import nem.parser
import nem.symbol_table
import nem.token_
import nem.types_
