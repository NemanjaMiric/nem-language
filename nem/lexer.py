"""Hold Lexer class.

Holds the Lexer class which is used for converting raw source code into tokens.

Tokens (Extended Backus-Naur form):
    LETTER_           = "A" | "B" | "C" | "D" | "E" | "F" | "G"
                      | "H" | "I" | "J" | "K" | "L" | "M" | "N"
                      | "O" | "P" | "Q" | "R" | "S" | "T" | "U"
                      | "V" | "W" | "X" | "Y" | "Z" | "a" | "b"
                      | "c" | "d" | "e" | "f" | "g" | "h" | "i"
                      | "j" | "k" | "l" | "m" | "n" | "o" | "p"
                      | "q" | "r" | "s" | "t" | "u" | "v" | "w"
                      | "x" | "y" | "z" ;
    DIGIT_            = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;
    ESCAPED_          = "\n" | "\t" | "\\" | '\"' ;
    NUMBER            = { DIGIT_ }, [ "." ], DIGIT_, { DIGIT_ } ;
    TEXT              = '"', { ? any character other than double quotation mark ? | ESCAPED_ }, '"' ;
    SYMBOL            = LETTER_ | "_", { LETTER_ | "_" | DIGIT_ } ;
    NULL              = "null" ;
    COMMA             = "," ;
    CARET             = "^" ;
    ASTERISK          = "*" ;
    SLASH             = "/" ;
    PLUS              = "+" ;
    MINUS             = "-" ;
    EQUAL             = "=" ;
    WHILE             = "while" ;
    LEFT_BRACKET      = "(" ;
    RIGHT_BRACKET     = ")" ;
    IF                = "if" ;
    OTHERWISE         = "otherwise" ;
    FUNCTION          = "function" ;
    IMPORT            = "import" ;
    LEFT_SQUARE       = "[" ;
    RIGHT_SQUARE      = "]" ;
    BREAK             = "break" ;
    CONTINUE          = "continue" ;
    RETURN            = "return" ;
    NOT               = "not" ;
    OR                = "or" ;
    AND               = "and" ;
    LESS              = "<" ;
    LESS_EQUAL        = "<=" ;
    GREATER           = ">" ;
    GREATER_EQUAL     = ">=" ;
    IS                = "is" ;
    MODULO            = "%" ;
    EOF               = ? end of file ? ;
    COMMENT_          = "#", { ? any character other than newline ? }, "\n" ;

"""

from nem.token_ import Token
from nem.exceptions import LexerException
import re


class Lexer:

    """Convert source code into tokens.

    Used for converting raw source code into tokens.

    """

    SYMBOLS = {
        "null": Token.NULL,
        "while": Token.WHILE,
        "if": Token.IF,
        "otherwise": Token.OTHERWISE,
        "function": Token.FUNCTION,
        "import": Token.IMPORT,
        "break": Token.BREAK,
        "continue": Token.CONTINUE,
        "return": Token.RETURN,
        "not": Token.NOT,
        "or": Token.OR,
        "and": Token.AND,
        "is": Token.IS
    }

    def __init__(self, code, filename):
        """Initialize Lexer class."""
        self.code = code
        self.filename = filename

        self.current_index = 0
        self.previous_character = None
        self.current_character = code[0]

        self.line = 1

    def _advance_index(self):
        """Advance the current character."""
        self.current_index += 1
        self.previous_character = self.current_character
        self.current_character = self.code[self.current_index] if self.current_index < len(self.code) else None
        if self.previous_character == "\n" and self.current_character is not None:
            self.line += 1

    def _next_number(self):
        """Return next number relative to the current position."""
        # Tries to find the next number
        while self.current_character is not None and not re.match("[0-9.]", self.current_character):
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No number found.".format(self.filename, self.line))

        length = 0

        # "[0-9]*\.?[0-9]*\Z" checks for a valid number (10, 10.2, .4, etc.)
        while self.current_character is not None                                                                       \
                and re.match("[0-9]*\.?[0-9]+\Z", self.code[self.current_index - length:self.current_index + 1])       \
                or self.current_character == ".":
            length += 1
            self._advance_index()

        if self.previous_character == ".":
            raise LexerException("Lexing Error (File {}) (Line {}): Unrecognized token"
                                 .format(self.filename, self.line))

        return self.code[self.current_index - length:self.current_index]

    def _next_symbol(self):
        """Return next symbol."""
        # Tries to find the next symbol
        while self.current_character is not None and not re.match("[a-zA-Z_]", self.current_character):
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No symbol found.".format(self.filename, self.line))

        length = 0

        # "[a-zA-Z_][a-zA-Z_0-9]*\Z" checks for a valid symbol, but many other tokens (purposefully) pass as well
        while self.current_character is not None                                                                       \
                and re.match("[a-zA-Z_][a-zA-Z_0-9]*\Z", self.code[self.current_index - length:self.current_index + 1]):
            length += 1
            self._advance_index()

        return self.code[self.current_index - length:self.current_index]

    def _next_text(self):
        """Return next text.

        Everything between the current or found quotation mark and the next unescaped one gets lexed.

        """
        # Tries to find the next text
        while self.current_character is not None and self.current_character != "\"":
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No text found.".format(self.filename, self.line))

        length = 0
        self._advance_index()

        escaped_characters = (
            "n",
            "t",
            "\"",
            "\\"
        )

        # Up until it reaches the closing quotation mark, it continues adding all the characters to text
        while self.current_character is not None and self.current_character != "\"":
            length += 1
            self._advance_index()

            if self.previous_character == "\\" and self.current_character in escaped_characters:
                length += 1
                self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): Text doesn't have an end"
                                 .format(self.filename, self.line))

        self._advance_index()

        # -1s are added because of self._advance_index() at the end, which increases self.current_index by 1
        return self.code[self.current_index - length - 1:self.current_index - 1]\
            .replace("\\n", "\n").replace("\\t", "\t").replace('\\"', "\"")

    def _next_token(self):
        """Return next token."""
        if self.current_character is None:
            token = None

        # Ignores whitespaces and tabs
        elif self.current_character in " \t":
            self._advance_index()
            token = False

        # Lexes COMMA token
        elif self.current_character == ",":
            token = Token(Token.COMMA, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes CARET token
        elif self.current_character == "^":
            token = Token(Token.CARET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes ASTERISK token
        elif self.current_character == "*":
            token = Token(Token.ASTERISK, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes SLASH token
        elif self.current_character == "/":
            token = Token(Token.SLASH, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes PLUS token
        elif self.current_character == "+":
            token = Token(Token.PLUS, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes MINUS token
        elif self.current_character == "-":
            token = Token(Token.MINUS, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes MODULO token
        elif self.current_character == "%":
            token = Token(Token.MODULO, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes EQUAL token
        elif self.current_character == "=":
            token = Token(Token.EQUAL, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LEFT_BRACKET token
        elif self.current_character == "(":
            token = Token(Token.LEFT_BRACKET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes RIGHT_BRACKET token
        elif self.current_character == ")":
            token = Token(Token.RIGHT_BRACKET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LEFT_SQUARE token
        elif self.current_character == "[":
            token = Token(Token.LEFT_SQUARE, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes RIGHT_SQUARE token
        elif self.current_character == "]":
            token = Token(Token.RIGHT_SQUARE, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LESS or LESS_EQUAL token
        elif self.current_character == "<":
            self._advance_index()
            if self.current_character == "=":
                token = Token(Token.LESS_EQUAL, "{}{}".format(self.previous_character, self.current_character),
                              self.filename, self.line)
                self._advance_index()
            else:
                token = Token(Token.LESS, self.previous_character, self.filename, self.line)

        # Lexes GREATER or GREATER_EQUAL token
        elif self.current_character == ">":
            self._advance_index()
            if self.current_character == "=":
                token = Token(Token.GREATER_EQUAL, "{}{}".format(self.previous_character, self.current_character),
                              self.filename, self.line)
                self._advance_index()
            else:
                token = Token(Token.GREATER, self.previous_character, self.filename, self.line)

        # Lexes TEXT token - everything between the current quotation mark and the next one gets lexed
        elif self.current_character == "\"":
            token = Token(Token.TEXT, self._next_text(), self.filename, self.line)

        # Lexes NUMBER token
        elif re.match("[0-9.]", self.current_character):
            token = Token(Token.NUMBER, self._next_number(), self.filename, self.line)

        # Lexes both SYMBOL tokens and tokens comprised of letters
        elif re.match("[a-zA-Z_]", self.current_character):
            temporary = self._next_symbol()

            try:
                token = Token(self.SYMBOLS[temporary], temporary, self.filename, self.line)
            except KeyError:
                # If it's not a token comprised of letters, it's a SYMBOL token
                token = Token(Token.SYMBOL, temporary, self.filename, self.line)

        elif self.current_character == "\n":
            self._advance_index()
            token = False

        # Comments - everything in-between '#' character and the next newline character gets ignored
        elif self.current_character == "#":
            while self.current_character is not None and self.current_character != "\n":
                self._advance_index()
            self._advance_index()

            token = False

        else:
            raise LexerException("Lexing Error (File {}) (Line {}): Unrecognized token"
                                 .format(self.filename, self.line))

        return token

    def _tokens(self):
        """Return all tokens."""
        current_token = self._next_token()

        while current_token is not None:
            # If the current token is a comment, skips the yield
            if current_token:
                yield current_token
            current_token = self._next_token()

        # Appends the EOF token
        yield Token(Token.EOF, None, self.filename, self.line)

    def lex(self):
        """Lex the code.

        Lexer for Nem code.

        Tokens (Extended Backus-Naur form):
            LETTER_           = "A" | "B" | "C" | "D" | "E" | "F" | "G"
                              | "H" | "I" | "J" | "K" | "L" | "M" | "N"
                              | "O" | "P" | "Q" | "R" | "S" | "T" | "U"
                              | "V" | "W" | "X" | "Y" | "Z" | "a" | "b"
                              | "c" | "d" | "e" | "f" | "g" | "h" | "i"
                              | "j" | "k" | "l" | "m" | "n" | "o" | "p"
                              | "q" | "r" | "s" | "t" | "u" | "v" | "w"
                              | "x" | "y" | "z" ;
            DIGIT_            = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;
            ESCAPED_          = "\n" | "\t" | "\\" | '\"' ;
            NUMBER            = { DIGIT_ }, [ "." ], DIGIT_, { DIGIT_ } ;
            TEXT              = '"', { ? any character other than double quotation mark ? | ESCAPED_ }, '"' ;
            SYMBOL            = LETTER_ | "_", { LETTER_ | "_" | DIGIT_ } ;
            NULL              = "null" ;
            COMMA             = "," ;
            CARET             = "^" ;
            ASTERISK          = "*" ;
            SLASH             = "/" ;
            PLUS              = "+" ;
            MINUS             = "-" ;
            EQUAL             = "=" ;
            WHILE             = "while" ;
            LEFT_BRACKET      = "(" ;
            RIGHT_BRACKET     = ")" ;
            IF                = "if" ;
            OTHERWISE         = "otherwise" ;
            FUNCTION          = "function" ;
            IMPORT            = "import" ;
            LEFT_SQUARE       = "[" ;
            RIGHT_SQUARE      = "]" ;
            BREAK             = "break" ;
            CONTINUE          = "continue" ;
            RETURN            = "return" ;
            NOT               = "not" ;
            OR                = "or" ;
            AND               = "and" ;
            LESS              = "<" ;
            LESS_EQUAL        = "<=" ;
            GREATER           = ">" ;
            GREATER_EQUAL     = ">=" ;
            IS                = "is" ;
            MODULO            = "%" ;
            EOF               = ? end of file ? ;
            COMMENT_          = "#", { ? any character other than newline ? }, "\n" ;

        """
        return self._tokens()


def main():
    """Debug the lexer.

    Used as the entry-point when the file gets ran directly. Used for debugging the class Lexer.

    """
    while True:
        print(list(Lexer(input(">> ") + "\n", "<stdin>").lex()))


if __name__ == "__main__":
    # Only activates when the file gets ran directly.
    main()
