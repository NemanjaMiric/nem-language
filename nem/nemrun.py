#!/usr/bin/env python
"""Implement nemrun command."""

import nem

import sys


def script():
    """Interpret nem file."""
    with open(sys.argv[1]) as code:
        # Built-in variables and functions
        symbol_table = nem.symbol_table.SymbolTable()
        symbol_table.set("true", nem.types_.Number(1))
        symbol_table.set("false", nem.types_.Number(0))
        symbol_table.set("print", nem.types_.BuiltInFunction(["element"], 0))
        symbol_table.set("input", nem.types_.BuiltInFunction([], 1))
        symbol_table.set("convert", nem.types_.BuiltInFunction(["value", "type"], 2))

        nem.interpreter.Interpreter(code.read(), sys.argv[1], symbol_table)
