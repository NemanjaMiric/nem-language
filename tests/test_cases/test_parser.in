# Fibonacci numbers generator

import "test"

function fib(first, second, stop) (
    while (not 0 and not 0) (
        if(first < stop) (
            temporary = first
            first = second
            second = temporary + second
        ) otherwise break
        continue
    )

    return 0
)

fib(1, 1, 1000)

a = [1, null, "\"text\""]
_0 = (- 0.2 + .2) ^ (2 + 2) * (-2 + 02) ^ (2 - 2) + (-2 - 2) ^ (2 + 2) * (2 + 2) ^ (2 + 2) # comments are cool

b = a[1]