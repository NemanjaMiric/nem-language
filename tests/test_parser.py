"""Test for Parser class.

Unit testing for the Parser class.

"""


import unittest
from nem.lexer import Lexer
from nem.parser import Parser
from nem.exceptions import ParserException
from nem.nodes import *


class ParserTestCase(unittest.TestCase):

    """Unit test Parser class.

    Used for unit testing the Parser class.

    """

    def test_parse(self):
        """Test the parse method.

        Tests the parser.

        """
        with open("test_cases/test_parser.in") as _input:
            _input = _input.read()

        with open("test_cases/test_parser.out") as output:
            output = output.read()

        self.assertEqual("".join(list(map(repr, Parser(Lexer(_input, "<stdin>").lex()).parse()))), output)

        with self.assertRaises(ParserException):
            list(Parser(Lexer("\n2+", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("import\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("nem(1, 2\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("function nem\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("function nem(a,b\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("function nem(a,\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("function nem(1\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("[1,2,3\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("a[1\n", "<stdin>").lex()).parse())

        with self.assertRaises(ParserException):
            list(Parser(Lexer("import\n", "<stdin>").lex()).parse())


if __name__ == '__main__':
    # Only activates when the file gets ran directly.
    unittest.main()
