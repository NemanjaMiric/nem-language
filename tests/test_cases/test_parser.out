Import('test')FunctionDefinition('fib', ['first', 'second', 'stop'], Expressions([While(Expressions([BinaryOperation(UnaryOperation(Number('0'), 'not'), 'and', UnaryOperation(Number('0'), 'not'))]), Expressions([IfOtherwise(Expressions([BinaryOperation(Variable('first'), '<', Variable('stop'))]), Expressions([AssignmentOperation('temporary', Variable('first')), AssignmentOperation('first', Variable('second')), AssignmentOperation('second', BinaryOperation(Variable('temporary'), '+', Variable('second')))]), Break()), Continue()])), Return(Number('0'))]))FunctionCall('fib', [Number('1'), Number('1'), Number('1000')])AssignmentOperation('a1', List([Number('1'), Null(), Text('"text"')]))AssignmentOperation('_0', BinaryOperation(BinaryOperation(BinaryOperation(Expressions([BinaryOperation(UnaryOperation(Number('0.2'), '-'), '+', Number('.2'))]), '^', Expressions([BinaryOperation(Number('2'), '%', Number('2'))])), '*', BinaryOperation(Expressions([BinaryOperation(UnaryOperation(Number('2'), '-'), '+', Number('02'))]), '^', Expressions([BinaryOperation(Number('2'), '-', Number('2'))]))), '+', BinaryOperation(BinaryOperation(Expressions([BinaryOperation(UnaryOperation(Number('2'), '-'), '-', Number('2'))]), '^', Expressions([BinaryOperation(Number('2'), '+', Number('2'))])), '/', BinaryOperation(Expressions([BinaryOperation(Number('2'), '+', Number('2'))]), '^', Expressions([BinaryOperation(Number('2'), '+', Number('2'))])))))AssignmentOperation('b', ListIndex(Variable('a'), Number('1')))FunctionDefinition('a', [], Expressions([FunctionDefinition('b', [], Return(List([Number('1'), List([Number('1'), Number('2'), List([Number('2'), Number('3')])])]))), Return(Variable('b'))]))AssignmentOperation('d', FunctionCall('convert', [ListIndex(ListIndex(ListIndex(FunctionCall(FunctionCall('a', []), []), Number('1')), Number('2')), Number('0')), Text('number')]))FunctionCall('print', [ListIndex(List([Number('1'), Number('2')]), Number('0'))])FunctionCall('print', [ListIndex(ListIndex(Text('asd'), Number('1')), Number('0'))])